### PR TITLE
`General`: Improve the handling of file extensions

### PR DESCRIPTION
**Pull request by ge65cer**
#### General

- [x] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.ase.in.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/development-process.html#naming-conventions-for-github-pull-requests).
#### Server
- [x] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/server/).
#### Client
- [x] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client/) and ensured that the layout is responsive.
- [x] Following the [theming guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client-design/), I specified colors only in the theming variable files and checked that the changes look consistent in both the light and the dark theme.
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client-tests/).
- [x] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.

### Motivation and Context

In #5328 support for new file extensions was added. However, the associated icons for attachment units were not updated. While working on that issue, I found some more potential for improvements.

### Description

Foremost, this PR adds icons for the new file extensions in attachment units.
It also allows users to upload some more file types: `gif, tar, txt, md, htm, html, json, odc, odi`.
Some other improvements include the removal of code duplication in `file-uploader.service.ts` and displaying the allowed file types also on the _lecture attachments_ form (previously only displayed in the _attachment units_ form).
Finally, the markdown editor has its own list of allowed file types (only images and PDFs), this restriction is now also enforced server-side.

### Steps for Testing

Prerequisites:
- 1 Instructor
- 1 Lecture

1. Log in to Artemis
2. Navigate to Course Administration
3. Create some _attachment units_ with different file types for the lecture
4. Try to submit the form without selecting a file or selecting disallowed file types
5. Take a look at the created units and ensure a corresponding icon is shown for the file type
6. Upload some _attachments_ for the lecture (should show allowed file types in tooltip)
7. Try to submit the form without selecting a file or selecting disallowed file types
8. Ensure uploading files to the markdown editor still works (e.g., in the exercise problem description)
9. Also try to upload disallowed file types to the markdown editor

#### Code Review
- [x] Review 1
- [x] Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

### Test Coverage

| Class/File | Branches | Lines/Instructions |
|------------|-------:|-----:|
| lecture-attachments.component.ts | 79% | 97% |
| attachment-unit-form.component.ts | 78% | 91% |
| FileResource.java | 78% | 92% |

### Screenshots

Screenshot 2022-08-08 at 13 35 39

Screenshot 2022-08-08 at 13 22 57Screenshot 2022-08-08 at 13 24 07

### CODE REVIEW

**Review by joaquinmurcia — APPROVED** [review]
Tested on TS2.

5. How are we supposed to create a code unit if code file formats (e.g. *.py) is not allowed in the attachment files?

7. When uploading an attachment to a lecture, trying to add it without submitting a file doesn't work, but it also doesn't display any error.

The rest of the functionality works as expected.

**Review by pantageepapa — COMMENTED** [review]
Tested on ts1 and it works as expected - however, there are more icons for different types than in the screenshot (e.g. for .csv and .pptx there are different icons) - is it as expected?

**Comment by ge65cer** [comment]
@pantageepapa This is correct, the screenshot is just exemplary.

**Review by bassner — CHANGES_REQUESTED** [review]
When I try to submit an invalid file or keep the title empty when uploading an attachment, nothing happens. It should disable the button and mark some of the form fields red or something when I can't submit for some reason, but not just fail silently.
(Basically testing step 7 fails / doesn't work correctly)

https://user-images.githubusercontent.com/23171488/184007659-c4d31d17-b9f6-40c7-a72d-54b81c2b55c2.mov

If you think this is not in scope of this PR as it was like this before (afaik), we could defer this to a follow up - just dismiss this review in that case.

**Review by Mikael31 — APPROVED** [review]
Tested on ts2. It showed all the corresponding icons, when i uploaded them.

**Comment by krusche** [comment]
> Finally, the markdown editor kept track of its own allowed file types (png, jpg, jpeg, svg, pdf) -- this has been replaced with the global list of allowed file extensions

I don't think it is possible to add other files than images into the markdown editor. Please make sure that you do not introduce usability or security issues by removing this.

**Comment by ge65cer** [comment]
@bassner Thanks for your review! I checked the current production build, you are correct that this bug was already there.
I have implemented a small fix that will disable the submit button for such cases without touching the rest of the component logic.

@krusche Okay, then let's keep the restriction to images & PDFs. However, for security reasons, we should then also enforce this limit in the server as well.

**Comment by krusche** [comment]
> @krusche Okay, then let's keep the restriction to images & PDFs. However, for security reasons, we should then also enforce this limit in the server as well.

I agree. If this is not enforced yet, we should do that

File: src/main/java/de/tum/in/www1/artemis/web/rest/FileResource.java [hunk 1]
```diff
@@ -557,32 +558,36 @@ private boolean validateTemporaryAccessTokenForCourse(String temporaryAccessToke
...
-            return ResponseEntity.badRequest().body("Unsupported file type! Allowed file types: " + String.join(", ", this.allowedFileExtensions));
-        }
 
+        final String fileExtension = FilenameUtils.getExtension(filename);
         final String filePath;
         final String fileNameAddition;
         final StringBuilder responsePath = new StringBuilder();
 
-        // set the appropriate values depending on the use case
+        // Check the allowed file extensions and set the appropriate values depending on the use case
```
Comment (b-fein) [inline src/main/java/de/tum/in/www1/artemis/web/rest/FileResource.java:576]: The duplication of the extension checking should be moved to its own method.

```java
private void checkAllowedFileExtensionElseThrow() {
    final Set<String> allowedExtensions;
    if (markdown) {
       allowedExtensions = allowedMarkdownFileExtensions;
    } else {
        allowedExtensions = allowedFileExtensions;
    }

    if (allowedExtensions.stream().noneMatch(fileExtension::equalsIgnoreCase)) {
        return ResponseEntity.badRequest().body("Unsupported file type! Allowed file types: " + String.join(", ", allowedExtensions));
    }
}
```

File: src/main/webapp/app/shared/http/file-uploader.service.ts [hunk 2]
```diff
@@ -14,79 +14,51 @@ type Options = {
...
+     * @param file The file to upload
+     * @param fileName The name of the file
+     * @param options The options dictionary (e.g, { keepFileName: true })
+     * @return A promise with the response from the server or an error
+     * @private
+     */
+    private handleFileUpload(endpoint: string, allowedExtensions: string[], file: Blob | File, fileName?: string, options?: Options): Promise<FileUploadResponse> {
         const fileExtension = fileName ? fileName.split('.').pop()!.toLocaleLowerCase() : file['name'].split('.').pop().toLocaleLowerCase();
-        if (this.acceptedFileExtensions.split(',').indexOf(fileExtension) === -1) {
+        if (allowedExtensions.indexOf(fileExtension) === -1) {
```
Comment (b-fein) [inline src/main/webapp/app/shared/http/file-uploader.service.ts:48]:
```suggestion
        if (allowedExtensions.includes(fileExtension)) {
```

File: src/main/webapp/app/shared/markdown-editor/markdown-editor.component.ts [hunk 3]
```diff
@@ -457,33 +456,25 @@ export class MarkdownEditorComponent implements AfterViewInit {
...
-                );
-            }
+            this.fileUploaderService.uploadMarkdownFile(file).then(
+                (res) => {
+                    const extension = file.name.split('.').pop()!.toLocaleLowerCase();
+
+                    let textToAdd = `[${file.name}](${res.path})\n`;
+                    if (extension !== 'pdf') {
+                        textToAdd = '!' + textToAdd;
+                    }
```
Comment (b-fein) [inline src/main/webapp/app/shared/markdown-editor/markdown-editor.component.ts:466]:
```suggestion
                    if (extension !== 'pdf') {
                        // show file as embedded images
                        textToAdd = '!' + textToAdd;
                    }
```

**Review by b-fein — CHANGES_REQUESTED** [review]
## Code Review
- change request: code duplication on the server that should be resolved
- two small suggestions for the client code

## Manual Testing
Lecture Attachments and files in the Markdown editor worked as expected, even ones with tricky names (no extension, multiple extensions, only extension: `.png`, with name `...`).

File: src/main/java/de/tum/in/www1/artemis/web/rest/FileResource.java [hunk 4]
```diff
@@ -557,26 +558,36 @@ private boolean validateTemporaryAccessTokenForCourse(String temporaryAccessToke
...
+        if (markdown) {
+            allowedExtensions = allowedMarkdownFileExtensions;
+        }
+        else {
+            allowedExtensions = allowedFileExtensions;
+        }
+
+        if (allowedExtensions.stream().noneMatch(fileExtension::equalsIgnoreCase)) {
+            return ResponseEntity.badRequest().body("Unsupported file type! Allowed file types: " + String.join(", ", allowedExtensions));
         }
```
Comment (b-fein) [inline src/main/java/de/tum/in/www1/artemis/web/rest/FileResource.java:584]: Thanks, that’s better than before. This method was already very long before the addition. That is why I suggested moving the check to a separate method, though.

Comment (ge65cer) [reply src/main/java/de/tum/in/www1/artemis/web/rest/FileResource.java:584]: I'd rather keep it as is, but happy to discuss if we get a third opinion.

**Review by bassner — APPROVED** [review]
Thanks for adding the changes. Looks good now.
I suggest that you talk to @julian-christl about potential conflicts with #5427 where he touched the same code

File: src/main/java/de/tum/in/www1/artemis/web/rest/FileResource.java [hunk 5]
```diff
@@ -557,26 +558,36 @@ private boolean validateTemporaryAccessTokenForCourse(String temporaryAccessToke
...
+        final Set<String> allowedExtensions;
+
+        if (markdown) {
+            allowedExtensions = allowedMarkdownFileExtensions;
+        }
+        else {
+            allowedExtensions = allowedFileExtensions;
+        }
+
+        if (allowedExtensions.stream().noneMatch(fileExtension::equalsIgnoreCase)) {
```
Comment (JohannesStoehr) [inline src/main/java/de/tum/in/www1/artemis/web/rest/FileResource.java:582]:
```suggestion
        if (!allowedExtensions.contains(fileExtension.toLowerCase())) {
```
This seams more intuitive to me

**Review by JohannesStoehr — CHANGES_REQUESTED** [review]
One small code nitpick:

File: src/main/java/de/tum/in/www1/artemis/web/rest/FileResource.java (hunk 5 above)
Comment (manuelmanso) [reply src/main/java/de/tum/in/www1/artemis/web/rest/FileResource.java:582]: Agreed 👍

**Review by manuelmanso — APPROVED** [review]
Changes look good (although I think the PR should have been split in two - one for the refactoring and improving the handling of the extensions and one for adding the new icons)

File: src/main/webapp/app/lecture/lecture-unit/lecture-unit-management/attachment-unit-form/attachment-unit-form.component.html [hunk 6]
```diff
@@ -36,8 +36,11 @@
...
-                <small id="fileHelp" class="form-text text-muted">{{ 'artemisApp.attachmentUnit.createAttachmentUnit.fileFormat' | artemisTranslate }} </small>
-                <fa-icon [icon]="farQuestionCircle" [ngbTooltip]="fileLimitations"></fa-icon>
+                <small id="fileHelp" class="form-text text-muted">{{ 'artemisApp.attachmentUnit.createAttachmentUnit.fileLimitation' | artemisTranslate }} </small>
+                <fa-icon
+                    [icon]="faQuestionCircle"
+                    [ngbTooltip]="'artemisApp.attachmentUnit.createAttachmentUnit.fileLimitationTooltip' | artemisTranslate: { extensions: allowedFileExtensions }"
+                ></fa-icon>
                 <div class="input-group background-file">
                     <div class="custom-file overflow-ellipsis">
                         <input
```
Comment (akesfeden) [inline src/main/webapp/app/lecture/lecture-unit/lecture-unit-management/attachment-unit-form/attachment-unit-form.component.html:46]: Adding [`accept`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/file#accept) attribute like the example below can further increase the usability by filtering out the ineligible file formats from the browser's file upload UI.  `allowedFileExtensions` property is almost in the correct format for `accept` but it needs dots before the extensions.
```
<input accept=".png,.jpg,.jpeg,.gif" ... >
```

File: src/main/webapp/app/lecture/lecture-attachments.component.html [hunk 7]
```diff
@@ -107,13 +107,19 @@ <h4 jhiTranslate="artemisApp.lecture.attachments.editAttachment" *ngIf="attachme
...
                         <div class="form-group">
                             <label class="form-control-label" jhiTranslate="artemisApp.lecture.attachments.file">File</label>
+                            <small id="fileHelp" class="form-text text-muted">{{ 'artemisApp.lecture.attachments.fileLimitation' | artemisTranslate }} </small>
+                            <fa-icon
+                                [icon]="faQuestionCircle"
+                                [ngbTooltip]="'artemisApp.lecture.attachments.fileLimitationTooltip' | artemisTranslate: { extensions: allowedFileExtensions }"
+                            ></fa-icon>
                             <div class="input-group background-file">
                                 <div class="custom-file overflow-ellipsis">
                                     <input
```
Comment (akesfeden) [inline src/main/webapp/app/lecture/lecture-attachments.component.html:117]: Same as [above](https://github.com/ls1intum/Artemis/pull/5578/files#r953065232)

File: src/main/webapp/app/lecture/lecture-unit/lecture-unit-management/attachment-unit-form/attachment-unit-form.component.html (hunk 6 above)
Comment (ge65cer) [reply src/main/webapp/app/lecture/lecture-unit/lecture-unit-management/attachment-unit-form/attachment-unit-form.component.html:46]: Good idea!

File: src/main/webapp/app/lecture/lecture-attachments.component.ts [hunk 8]
```diff
@@ -29,7 +29,8 @@ export class LectureAttachmentsComponent implements OnInit, OnDestroy {
     erroredFile?: Blob;
     errorMessage?: string;
 
-    readonly allowedFileExtensions = FILE_EXTENSIONS.join(', ');
+    readonly allowedFileExtensions = FILE_EXTENSIONS.map((ext) => '.' + ext).join(',');
+    readonly allowedFileExtensionsString = FILE_EXTENSIONS.join(', ');
```
Comment (b-fein) [inline src/main/webapp/app/lecture/lecture-attachments.component.ts:33]: I don’t think `…String` is a good name. Maybe `acceptedFileExtensionsFileBrowser` and `allowedFileExtensions` with a comment before the first one why two separate ones are needed would be better?

**Review by akesfeden — COMMENTED** [review]
Tested on TS1, mostly works well but when I upload 2 attachments to the lecture (not as a unit) with the same file name and then delete one of them, other file also becomes inaccessible. I am not sure if that is directly related to this PR though.

**Comment by ge65cer** [comment]
@akesfeden That bug is already known and tracked in #4937 (not related to this PR).

**Review by akesfeden — APPROVED** [review]
Thanks for the explanation, then I haven't noticed any other issues during my tests on TS1, invalid extensions are rejected and icons are shown as expected.

**Review by JohannesStoehr — APPROVED** [review]
Code looks good to me

**Review by maximiliansoelch — APPROVED** [review]
Tested on TS3, works as expected 👍

**Review by Hialus — APPROVED** [review]
Code changes look good to me.

**Pull request merged**